### PR TITLE
feat: format_test

### PR DESCRIPTION
**Pull request by hunshcn**
Resolve https://github.com/aspect-build/rules_lint/issues/175

---

### Type of change

- New feature or functionality (change which adds functionality)
- Documentation (updates to documentation or READMEs)

**For changes visible to end-users**

- Relevant documentation has been updated
- Suggested release notes are provided below:
```
new rule `format_test` is a available to create format test. see docs/format.md for more infomation.
```

### Test plan

- New test cases added
- Manual testing; please provide instructions so we can reproduce:
Because the wrong format file cannot be imported into the repo, the failure test of `workspace mode` can only be performed manually. (Manually change some code to the wrong format.)

### CODE REVIEW

**Comment by hunshcn** [comment]
@alexeagle

**Comment by alexeagle** [comment]
Ugh, the test failure looks like something latent that this has tickled :(

**Comment by hunshcn** [comment]
C++_with_clang-format passed locally, let me open a codespace.

`Java_with_java-format` and `Kotlin_with_ktfmt` failed even at main branch in my m1 mac and arm64 linux. I have no idea why

**Comment by hunshcn** [comment]
in codespace, all passed.........(The swiftformat failed by so file is missing, which is reasonable.)

image

**Comment by hunshcn** [comment]
clang need libtinfo.so.5, so we need update ci and install this dep before test.

bazel-contrib/.github/.github/workflows/bazel.yaml@v5 does not support. v6 is required ( configurate bazel_test_command), but new problems happen.

Java is still broken, I don't know why, and I don't understand it.

**Comment by hunshcn** [comment]
if turn off clang, revert changes of ci.yaml

File: docs/formatting.md [hunk 1]
```diff
@@ -88,6 +88,26 @@ $ chmod u+x .git/hooks/pre-commit
...
+
+```starlark
+load("@aspect_rules_lint//format:defs.bzl", "format_test")
+
+format_test(
+    name = "format_test",
+    # register languages, e.g. same to format_multirun
+    # python = "//:ruff",
+    workspace = "//:WORKSPACE.bazel", # not cacheable
+    # srcs = ["files"], # cacheable, but need declare all files. recommend use with kazel
```
Comment (alexeagle) [inline docs/formatting.md:109]: note, I've never heard of `kazel` before, most Bazel users are familiar with Gazelle for build file generation. I'll update this.

Comment (hunshcn) [reply docs/formatting.md:109]: ok

**Comment by alexeagle** [comment]
Hm, I think we should probably have the same API as `buildifier_test` with https://github.com/bazelbuild/buildtools/pull/1092/files#diff-c5e2ba5db338e88cf7c850afd25c044dfa97198fefa4b296370c9286ab81d415R97 in addition to `srcs`/`workspace`, any thoughts about that @hunshcn ?

**Comment by hunshcn** [comment]
> Hm, I think we should probably have the same API as `buildifier_test` with a `no_sandbox` attribute https://github.com/bazelbuild/buildtools/pull/1092/files#diff-c5e2ba5db338e88cf7c850afd25c044dfa97198fefa4b296370c9286ab81d415R97 in addition to `srcs`/`workspace`, any thoughts about that @hunshcn ?

While providing workspace and set no_sandbox=True is strongly binding, this does provide clear semantics that I think are great.

**Comment by alexeagle** [comment]
Yeah I think it's just there to ensure that you read the documentation and know that the test will never be a cache hit. Will add.

**Comment by alexeagle** [comment]
Thanks for getting this together @hunshcn

**Review by alexeagle — APPROVED** [review]

**Comment by Zemnmez** [comment]
Thanks for this!

**Pull request merged**